Move array_poolt class to its own files

romainbrenguier opened the pull request:
This makes it easier to find array_poolt related code.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier committed:
Move array_poolt class to its own files

This makes it easier to find array_poolt related code.

tautschnig commented:
@romainbrenguier clang-format, cpplint, doxygen are unhappy.

owen-mc-diffblue commented on src/solvers/strings/array_pool.cpp:
```diff
@@ -0,0 +1,164 @@
+
+#include "array_pool.h"
+
+/// generate a new symbol expression of the given type with some prefix
```
Capitalise first word :bird:

owen-mc-diffblue commented on src/solvers/strings/array_pool.cpp:
```diff
@@ -0,0 +1,164 @@
+
+/// generate a new symbol expression of the given type with some prefix
+/// \par parameters: a prefix and a type
+/// \return a symbol of type tp whose name starts with "string_refinement#"
```
What is "tp"? Should it be replaced by "\p type"

owen-mc-diffblue commented on src/solvers/strings/array_pool.cpp:
```diff
@@ -0,0 +1,164 @@
+  return s.length();
+}
+
+/// construct a string expression whose length and content are new variables
```
:bird:

owen-mc-diffblue commented on src/solvers/strings/array_pool.cpp:
```diff
@@ -0,0 +1,164 @@
+  return array_pool.find(string_argument.op1(), string_argument.op0());
+}
+
+/// casts an expression to a string expression, or fetches the actual
```
:bird:

owen-mc-diffblue commented on src/solvers/strings/array_pool.cpp:
```diff
@@ -0,0 +1,164 @@
+  return *created_strings_value.insert(array).first;
+}
+
+/// Adds creates a new array if it does not already exists
```
Does it add it or create it?

owen-mc-diffblue commented on src/solvers/strings/array_pool.h:
```diff
@@ -0,0 +1,95 @@
+  fresh_string(const typet &index_type, const typet &char_type);
+
+private:
+  // associate arrays to char pointers
```
:bird:

owen-mc-diffblue replied:
`//` -> `///` to trigger doxygen :cat:

owen-mc-diffblue commented on src/solvers/strings/array_pool.h:
```diff
@@ -0,0 +1,95 @@
+  // associate arrays to char pointers
+  std::unordered_map<exprt, array_string_exprt, irep_hash> arrays_of_pointers;
+
+  // associate length to arrays of infinite size
```
:bird: :cat:

owen-mc-diffblue commented on src/solvers/strings/array_pool.h:
```diff
@@ -0,0 +1,95 @@
+  std::unordered_map<array_string_exprt, symbol_exprt, irep_hash>
+    length_of_array;
+
+  // generates fresh symbols
```
:bird: :cat:

owen-mc-diffblue commented on src/solvers/strings/array_pool.h:
```diff
@@ -0,0 +1,95 @@
+  // generates fresh symbols
+  symbol_generatort &fresh_symbol;
+
+  // Strings created in the pool
```
:bird: :cat:

owen-mc-diffblue reviewed (approved):
I've made some suggestions for cleaning up the comments a bit. I appreciate that you're just moving code, so feel free to ignore them if you don't have time.

romainbrenguier committed:
Update documentation of char_array_of_pointer

The old one was not clear, and should be moved to the header file since
it is part of the public interface.

romainbrenguier committed:
Update documentation of symbol_generatort

Fix some error and move it to the header.

romainbrenguier committed:
Update array_poolt documentation

Move documentation of public function to the header and clarify some
points.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 876c25d4cbadf1cc852322c6a21978268568b2cd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108499851
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 91349a12a110357b5e0bbf16398a493f22253a96).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108500135

tautschnig commented:
@romainbrenguier Could you please rebase? Travis has finished, but for some reason GitHub did not get a notification thereof.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4563c75626051c0706d8ef940de2f8639bcd1a27).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108656011
